### ISSUE TITLE
let-flow executes body without waiting for side-effects

### ISSUE THREAD

**Issue by yonatane** [ztellman/manifold#118]
The let-flow doc string says "The body will only be executed once all of the let-bound values, even ones only used for side effects, have been computed.", but when binding a value for side-effects only, without it appearing in the body, let-flow doesn't wait for it before executing the body. It doesn't even execute the side-effect.

Example (in REPL):
Not referencing 'late in the body:
```
(d/let-flow [immediate (d/success-deferred "immediate")
             late (d/future
                    (Thread/sleep 1000)
                    (println "late"))]
  (println immediate))
immediate
=> << nil >>
```
Referencing 'late in the body:
```
(d/let-flow [immediate (d/success-deferred "immediate")
             late (d/future
                    (Thread/sleep 1000)
                    (println "late"))]
  late
  (println immediate))
=> << … >>
late
immediate
```

**Comment by ztellman** [comment]
This is a clear disconnect between the documentation and implementation.  The current implementation intentionally does not execute anything which is not referenced elsewhere.  This has some nice properties, so I'm tempted to simply update the documentation to reflect the current behavior.

I apologize for the confusion here, but if you have any opinions on this I'm happy to take them into consideration.

**Comment by dm3** [comment]
@yonatane You can wait for *all* deferreds using `manifold.deferred/zip` and for the first of several via `alt` which you can copy from #102. Hopefully that one will get into master soon enough!

**Comment by skaplan** [comment]
The execution of the body will wait for side-effects if the side-effect bindings reference non-side-effect bindings:
```
(let-flow [a (do-work]
           _ (side-effect a)]
   a)
```
In this example, side-effect will be called.

I would prefer that let-flow does wait for side effects. Code written this way is more elegant.
```
(let-flow [data (get-data)
           _ (cache data)]
  data)
```
is better than
```
(let-flow [data (get-data)]
  (chain (cache data)
     (fn [_]
        data))
```
As the code-complexity grows, the difference between the two increases.

Considering that the dependencies are calculated at compile-time, I'm curious when it is useful to not execute all the supplied bindings, since if there was an unused binding, shouldn't it be removed in the code?

Regardless, I think we can get the best of both worlds by only executing side-effects when the binding is specifically marked using `_`. Would this be worth creating a pull request for?

**Comment by ztellman** [comment]
I can't claim an airtight rationale here, but the general notion of `let-flow` is "the action happens once all the necessary values are available".  Having the body block on data that it doesn't rely on would go against this general rule of thumb.

However, I acknowledge that this has been a recurring source of confusion, so empirically having the body wait on *everything* in the binding is more intuitive.  I think I'll just do that.

**Comment by aengelberg** [comment]
The fact that `let-flow` unpredictably omits entire forms really trips me up, because my use cases often involve side-effecting operations that the body does not depend on by name. I'd like to at least have the guarantee that all the code I write will get executed at some point.

I also think that by not making the body depend on all aforementioned forms by default, you allow the coder to unknowingly write "fire-and-forget" logic that has no backpressure, so something like

```clojure
(defn handle-request
  [ctx]
  (d/let-flow [foo (do-a-thing-to-a-database ...)
               _ (send-metric! (:id foo))
               bar (do-another-thing-i-guess ...)
               _ (send-metric! (:id bar))]
    response))
```

would leak memory if the `send-metric!` function suddenly started taking a long time.

**Comment by ztellman** [comment]
Sorry, I should have gotten to this ages ago.  I'll fix it and cut a release.

**Comment by minikomi** [comment]
Thanks @ztellman!
